### DISCUSSION TITLE
Improving Rails compatability

### DISCUSSION THREAD

**Discussion by wtn** [socketry/async-websocket#56] [General]
When using the [Rails adapter](https://github.com/socketry/async-websocket/blob/06ff17dc3bf2af9f5af1ca0ff1e0751b5dfb0446/lib/async/websocket/adapters/rails.rb), there are some side effects.

For example, there's no more `ActionDispatch::RequestId` inside the adapter block, so the Rails log entries after that point don't have a UUID:

```
[b6dd98ad-773c-4d0e-8071-c90e4aeab370] p0 fiber=41780 thread=41640
[b6dd98ad-773c-4d0e-8071-c90e4aeab370] Completed 101 Switching Protocols in 19ms (ActiveRecord: 4.1ms | Allocations: 7388)
[b6dd98ad-773c-4d0e-8071-c90e4aeab370] RackProtocolToConnectionUpgrade found rack.protocol=websocket in HTTP/1.1 response, setting 'upgrade' and 'connection' headers
p1 fiber=41780 thread=41640
p2 fiber=41780 thread=41640
```

For another example, if multiple databases are used with Active Record, any switching has to be re-done inside the adapter block.

**Comment by ioquatix** [comment]
Have you enabled fiber per request mode?

I'll have to take a look at `ActionDispatch::RequestId`, do you know if it uses fiber local storage or thread local storage?

https://blog.saeloun.com/2022/02/23/rails-fiber-safe-connection-pools/ has a good explanation.

(BTW, I don't know if this is the solution or not, maybe we need to use `Fiber.storage`.)
